src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MtAmazonDynamoDbStreamsBySharedTable.java

```diff
@@ -159,68 +187,20 @@ private int retryLoadRecords(MtGetRecordsResult mtResult,
     private static int addRecords(List<? super MtRecord> mtRecords,
                                   int limit,
                                   List<? extends Record> records,
-                                  Function<Record, MtRecord> recordMapper,
-                                  Predicate<MtRecord> recordFilter) {
+                                  Predicate<Record> recordFilter,
+                                  Function<Record, MtRecord> recordMapper) {
         int consumedRecords = 0;
         for (Record record : records) {
-            MtRecord mtRecord = recordMapper.apply(record);
-            if (recordFilter.test(mtRecord)) {
+            if (recordFilter.test(record)) {
                 if (mtRecords.size() >= limit) {
                     break;
                 } else {
-                    mtRecords.add(mtRecord);
+                    mtRecords.add(recordMapper.apply(record));
```

Comment:
This is the main optimization: instead of converting each record first and then filtering it, filter first and then convert (using a cached mapper rather than establishing tenant context for each record).